### PR TITLE
Use plug

### PR DESCRIPTION
**Pull request by josevalim**
I decided to use Plug for a more apple to apple comparison with other languages. Even though Phoenix overhead is low, it stills supports static assets, performs content negotiation, and more. But perhaps more importantly, it makes the application size between languages more comparable. :)

Something else that was also impacting the results is that you set the logger level to `:warning` in the console backend and not on `config :logger, level: :warning`. This means the logger machinery is invoked and the results are discarded just before writing to console. The logger machinery should be fast, but it adds a bit. In any case, this version does not add `Plug.Logger`, so it is no longer a concern.

Finally, I also added a configuration so Cowboy keeps connections alive for longer. IIRC it defaults to 100 requests, which is a sane default for actual apps, but probably too low for benchmarks.

If you want to squeeze a bit more, I would also suggest moving to [persistent_term](https://www.erlang.org/doc/man/persistent_term.html). If the data never or rarely changes (such as hourly), it will be cleaner and more performant than ETS.

PS: I could not download the transit data (I get a 403) but I am hoping the code still works. I did benchmark with an empty JSON response though and this version is roughly twice faster.

### CODE REVIEW

File: trexit/README.md [hunk 1]
```diff
@@ -1,5 +1,5 @@
 Run with
 
 ```
-> env MIX_ENV=prod SECRET_KEY_BASE="somerandomthingdoesntmatter" mix phx.server
+> env MIX_ENV=prod mix run --no-halt
```
Comment (josevalim) [inline trexit/README.md:4]:
```suggestion
> env MIX_ENV=prod mix run --preload-modules --no-halt
```

This makes sure all modules are loaded upfront. Add this flag if the benchmarking tool you are using does not perform warmup.

File: trexit/lib/trexit/endpoint.ex [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  plug :dispatch
+
+  get "/schedules/:route_id" do
+    json(conn, Trexit.GTFS.schedules_for_route(route_id))
+  end
+
+  defp json(conn, json) do
+    conn
+    |> put_resp_content_type("application/json")
+    |> send_resp(200, Jsonrs.encode_to_iodata!(json))
```
Comment (Qqwy) [inline trexit/lib/trexit/endpoint.ex:14]: Note that for Jsonrs, there is no difference between `encode!` and `encode_to_iodata!`.

Also, we could add the option `lean: true` which would disable usage of the custom struct encoding protocol (which we are not using) and be more efficient ([docs](https://hexdocs.pm/jsonrs/readme.html#quick-start)).

**Comment by rktjmp** [comment]
This patch gives a _small_ perf bump of around 525% with this patch applied.

```diff
<      data_received..................: 5.1 GB 92 MB/s
<      data_sent......................: 930 kB 17 kB/s
<      http_req_blocked...............: avg=16.73µs  min=783ns    med=2.33µs  max=7.92ms p(90)=3.42µs   p(95)=3.94µs 
<      http_req_connecting............: avg=14.37µs  min=0s       med=0s      max=7.87ms p(90)=0s       p(95)=0s     
<      http_req_duration..............: avg=265.54ms min=278.63µs med=81.43ms max=3.18s  p(90)=766.43ms p(95)=1.27s  
<        { expected_response:true }...: avg=265.51ms min=278.63µs med=81.43ms max=3.18s  p(90)=766.56ms p(95)=1.27s  
<      http_req_failed................: 0.01%  ✓ 1          ✗ 9899
<      http_req_receiving.............: avg=446.39µs min=8.72µs   med=89.52µs max=30.4ms p(90)=1.18ms   p(95)=2.06ms 
<      http_req_sending...............: avg=16.07µs  min=3.55µs   med=10.12µs max=9.07ms p(90)=14.62µs  p(95)=16.81µs
<      http_req_tls_handshaking.......: avg=0s       min=0s       med=0s      max=0s     p(90)=0s       p(95)=0s     
<      http_req_waiting...............: avg=265.08ms min=253.63µs med=81.2ms  max=3.17s  p(90)=765.85ms p(95)=1.27s  
<      http_reqs......................: 9900   177.986786/s
<      iteration_duration.............: avg=26.29s   min=703.36ms med=27.8s   max=48.22s p(90)=46.43s   p(95)=47.25s 
<      iterations.....................: 100    1.797846/s
<      vus............................: 11     min=11       max=50
<      vus_max........................: 50     min=50       max=50
---
>      data_received..................: 4.5 GB 84 MB/s
>      data_sent......................: 4.9 MB 91 kB/s
>      http_req_blocked...............: avg=2.43µs  min=572ns    med=1.02µs   max=6.8ms   p(90)=2.36µs  p(95)=2.86µs  
>      http_req_connecting............: avg=520ns   min=0s       med=0s       max=3.35ms  p(90)=0s      p(95)=0s      
>      http_req_duration..............: avg=42.51ms min=119.9µs  med=4.49ms   max=3.32s   p(90)=25.04ms p(95)=187.15ms
>        { expected_response:true }...: avg=42.5ms  min=119.9µs  med=4.49ms   max=3.32s   p(90)=25ms    p(95)=187.1ms 
>      http_req_failed................: 0.00%  ✓ 1          ✗ 51974
>      http_req_receiving.............: avg=96.64µs min=6.77µs   med=14.19µs  max=30.86ms p(90)=78.93µs p(95)=298.6µs 
>      http_req_sending...............: avg=12.94µs min=3.1µs    med=4.83µs   max=24.62ms p(90)=10.75µs p(95)=13.33µs 
>      http_req_tls_handshaking.......: avg=0s      min=0s       med=0s       max=0s      p(90)=0s      p(95)=0s      
>      http_req_waiting...............: avg=42.4ms  min=101.03µs med=4.46ms   max=3.32s   p(90)=24.87ms p(95)=186.66ms
>      http_reqs......................: 51975  973.704487/s
>      iteration_duration.............: avg=4.21s   min=350.92ms med=494.68ms max=41.21s  p(90)=24.22s  p(95)=30.37s  
>      iterations.....................: 525    9.835399/s
>      vus............................: 11     min=11       max=50 
>      vus_max........................: 50     min=50       max=50 

```

**Comment by losvedir** [comment]
Wow @josevalim! I'm 🤩 .

Thanks for the great ideas here. I just pushed up a commit fixing the logger level issue. I'll do a separate one for the cowboy keep-alive. I'm interested in how these parts individually contribute. Finally, I'll probably do another change for Plug, though I also want to look into a bit what the other languages are doing. .NET for example uses ASP.net, Scala is using Cask, and rust is using Axum. Only Go is using framework-less standard library, so I'm thinking maybe another approach would be to make sure Go is doing the things that a real webapp should be doing.

But I got tons of PRs and issues on this repo (thanks everyone!), and I just started a new job, so it's kind of slow going. I will get to it eventually, though.

**Comment by losvedir** [comment]
Okay, took a while but I finally had a chance to re-benchmark all my apps (since I updated to macOS Ventura), and then rebase this branch on the latest `main`, and re-benchmark Elixir again.

The performance was actually roughly comparable to the Phoenix version (with a slight edge towards Plug for the highest response rate one), which I think is a strong credit to Phoenix! But I figured I should still merge this since, as you say, it makes the code more apples to apples with the other languages.

I'll definitely look at `persistent_term`. I didn't think to use it since I never had to at work, but it actually makes a lot of sense for this use case (and I'll have to mention it to the folks who manage the _actual_ repo that this one is somewhat based on).

>  I could not download the transit data (I get a 403) but I am hoping the code still works.

Sigh, yeah the MBTA blocks traffic from a lot of countries for some misguided reason. 😢 

> I did benchmark with an empty JSON response though and this version is roughly twice faster.

Yeah, I didn't realize ahead of time, but the benchmark ends up being heavily influenced simply by the JSON encoding step. I got a 2x bump switching from `Jason` to `Jsonrs`, which surprised me since I've always used `Jason` (well, after the days of `Poison` anyway) without thinking twice.

**Pull request merged**